### ISSUE TITLE
Port Sound system to SFML

### ISSUE THREAD

**Issue by Quintus** [Secretchronicles/TSC#436]
The sound system needs to be ported from SDL to SFML.

See also http://www.sfml-dev.org/tutorials/2.3/audio-sounds.php

Valete,
Quintus

**Comment by Quintus** [comment]
@Luiji @datahead8888 @brianvanderburg2 this is one of the more important SFML porting issues, since a lot of objects in TSC emit sounds in some way or another. It would make porting of these objects smoother if a working sound system was in place. Please give a thought as to whether you want to tackle this issue.

The current audio system resides in the `src/audio` directory, but as it is closely interlined with the SDL soundsystem, you’re going to have to rewrite quite a bit of that. SFML has a nice audio tutorial available: http://www.sfml-dev.org/tutorials/2.3/audio-sounds.php

Valete,
Quintus

**Comment by kirbyfan64** [comment]
Ooooh, if I did this, could I add support for making songs loop without the intros?

**Comment by Quintus** [comment]
If you know how to tell SFML this, then yes. However, it is required that it also works with the current system, because we don’t want to do heavy systematic changes (besides SFML itself) in the SFML port. So if you want to tackle this, my suggestion is to first implement the basic system as we have it now. In a second step (and probably second PR) you expand the API of the sound system to support a new function `Play_With_Intro()` or similar that allows use of songs with intros.

I didn’t know you were a coder also, kirbyfan :-)

Vale,
Quintus

**Comment by datahead8888** [comment]
One of the folks in the sfml room suggested first converting all sfml while keeping legacy OpenGL intact and then tackling the legacy OpenGL.  I'm not immediately sure if this differs from the current strategy, but it might make sense to at least knock off all the SDL conversion tasks first, regardless.

Are we still planning the list of tasks, or will we start assigning to tasks?  

@kirbyfan64 we won't be able to directly assign you to a task because of github's limitations (you're not part of the core team yet).  We can note it in the ticket, though.  If you keep contributing as you have been, I have little doubt you will be part of the team soon.

**Comment by Quintus** [comment]
datahead8888 <notifications@github.com> writes:
> One of the folks in the sfml room suggested first converting all sfml
> while keeping legacy OpenGL intact and then tackling the legacy
> OpenGL.  I'm not immediately sure if this differs from the current
> strategy, but it might make sense to at least knock off all the SDL
> conversion tasks first, regardless.

We have surprisingly tiny amounts of raw OpenGL in our code. There is
much, much more SDL code. When I ported the contents of video.cpp
(sprite loading, manipulation, and rendering) to SFML’s sprite system, I
ported most of the raw OpenGL already. The part that currently has the
most legacy OpenGL that is still unported are for sure the particle
emitters (and the particle system in general).

> Are we still planning the list of tasks, or will we start assigning to
> tasks?

I have been opening a lot of SFML tickets, and I think I will continue
doing so a little. I will drop a notice when I think I have opened
everything that I deem worthwile. However, if you want to start working
on any of those tickets already, feel free to do so. I am the last who
will stop you :-)

> If you keep contributing as you
> have been, I have little doubt you will be part of the team soon.

Absolutely.

Vale,
Quintus

**Comment by datahead8888** [comment]
@kirbyfan64 - if you do this SFML conversion task, it should give you plenty of experience with the code for when you do the new feature for allowing intros (or endings) to be separated from a loopable portion of a song.  If you start this task, let us know, so that no one else does duplicated work.

**Comment by kirbyfan64** [comment]
So, this would basically be changing the source inside the `audio` directory to use SFML instead of SDL, right? I can definitely do that!

**Comment by Quintus** [comment]
Ryan Gonzalez <notifications@github.com> writes:

> So, this would basically be changing the source inside the `audio`
> directory to use SFML instead of SDL, right?

Yes.

> I can definitely do that!

Okay, so nobody else will work on this for now. When you are done with
your changes, please submit a PR (either by mail to the ML or with
GitHub’s web interface, depending on whether you prefer to work on
GitHub or not).

Vale,
Quintus

**Comment by kirbyfan64** [comment]
Slight oddity: SFML has no support for fading. I'll leave that for last!

Also, is the only purpose of channels in SDL mixer just like a global list of currently playing songs? Again, SFML doesn't have that concept.

**Comment by datahead8888** [comment]
> SFML has no support for fading. I'll leave that for last!

Per http://stackoverflow.com/questions/5498474/c-sfml-1-6-smooth-sound-fading-effects you might look into SetVolume() and SetAttenuation() functions within the API.  It looks like you may have to drive the fading logic yourself, though.

> Also, is the only purpose of channels in SDL mixer just like a global list of currently playing songs? Again, SFML doesn't have that concept.

This webpage seems to indicate that SFML supports channels:
http://www.sfml-dev.org/tutorials/1.6/audio-sound.php

My experience is pretty limited with SFML at the moment, so you might need to search into this a little more still.

**Comment by Quintus** [comment]
> My experience is pretty limited with SFML at the moment

You should take a day to read [the tutorials](http://www.sfml-dev.org/tutorials/2.3/). SFML’s API is so easy that I managed to get up and running with it in one day. The API reference also is very well done (a really nice example of a good API reference!).

Vale,
Quintus

**Comment by Quintus** [comment]
> http://www.sfml-dev.org/tutorials/1.6/audio-sound.php

This is an SFML 1.6 tutorial! Do not use that without knowing what you do. The current version is SFML 2.3.2, which is a breaking major release later! The audio stuff is now split over multiple tutorials, see the link I posted above.

Vale,
Quintus

**Comment by datahead8888** [comment]
@Quintus, do you think think SFML supports multiple sound channels?  This was @kirbyfan64 's question.  He may start work again before I read through all the tutorials.

What do you think on fading?

**Comment by kirbyfan64** [comment]
I think I may not need channels at all, simply because of the way the SFML API was designed. It is indeed much better than SDL!

**Comment by datahead8888** [comment]
@kirbyfan64 do we need channels in order to retain stereo sound?

**Comment by kirbyfan64** [comment]
Heads up: this may take a bit longer than I thought, since SFML's API, although MUCH better than SDL's mixer crud, is SUPER different. I had to change more than I realized!

**Comment by brianvanderburg2** [comment]
I'd like to help but probably will not be able to.  I still need to check out the branch, read some SFML documentation, brush up on some things.  Hopefully I'll get some time, even if only a few lines here or there or some comments/discussion.  Working 1 hr from home and on mandatory overtime, plus school, plus self-study for CCNP, plus real life is taking up all my time and more for the time being.  

As for the sound system, one thing that would be interesting is if SFML supports playing multiple sound tracks at a time.  If so, then perhaps support for fading, even if it must be done manually volume-wise, from one music track to another when it ends would be nice.  I think SDL only supported one music file at a time, or maybe that was just something with the SMC code.

**Comment by Quintus** [comment]
> As for the sound system, one thing that would be interesting is if SFML supports playing multiple sound tracks at a time.

I can’t infer that from the SFML documentation. I fear you simply have to test it by creating multiple instances of `sf::Music` and trying to play them simultaneously. Alternatively, you can join our IRC channel, because since recently we have an SFML dev in there (expl0iter). Same goes for you, @kirbyfan64 :-)

Valete,
Quintus

**Comment by datahead8888** [comment]
@kirbyfan64 -

> Also, is the only purpose of channels in SDL mixer just like a global list of currently playing songs? Again, SFML doesn't have that concept.

The 2.3 SFML documentation does talk about channels: http://www.sfml-dev.org/tutorials/2.3/audio-sounds.php

~~~~~~~~~~~~~~~~~~~~~~C++
std::vector<sf::Int16> samples = ...;
buffer.loadFromSamples(&samples[0], samples.size(), 2, 44100);
~~~~~~~~~~~~~~~~~~~~~~

> Since loadFromSamples loads a raw array of samples rather than an audio file, it requires additional arguments in order to have a complete description of the sound. The first one (third argument) is the number of channels; 1 channel defines a mono sound, 2 channels define a stereo sound, etc. The second additional attribute (fourth argument) is the sample rate; it defines how many samples must be played per second in order to reconstruct the original sound.

In audio.cpp in branch devel, in the constructor, I see this line:
~~~~~~~~~~~~~~~~~c++
m_audio_channels = MIX_DEFAULT_CHANNELS; // 1 = Mono, 2 = Stereo
~~~~~~~~~~~~~~~~~ 

This suggests that we need the channels in order to properly support stereo sound.  I'm not sure to what extent we're currently using stereo, but it would make sense to set the code up for it.

**Comment by kirbyfan64** [comment]
Status update: although there were some delays (*cough* schoolwork *cough*), it's going well. I ended up removing some completely unused APIs that would be hard to port to SFML.

Sorry about the delays. Hopefully I'll be finished by the weekend!

**Comment by datahead8888** [comment]
Thanks @kirbyfan64 - we really appreciate the help on the SFML port.  I definitely know the feeling with schoolwork burning up time.

**Comment by kirbyfan64** [comment]
Well, here's what I have so far: https://github.com/kirbyfan64/TSC/tree/sfml_audio. The diff is [HUGE](https://github.com/Secretchronicles/TSC/compare/devel...kirbyfan64:sfml_audio)!! I ended up removing a lot of code related to initializing, which SFML handles automatically.

Now, there's a slight issue. If one removes the cursor from the window, the music plays normally, but all the sprites freeze. However, with the cursor OVER the window, the sound stutters, and the sprites move normally.

Opening TSC in gdb shows that, when the mouse is over the window, hundreds of threads are being spawned and immediately exiting. Since SFML plays music on its own thread, I'm guessing this is the problem. Does anyone know why?

**Comment by Quintus** [comment]
Implemented with merge of #488. Closing.
